core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -617,6 +620,15 @@ public final NodeChannels openConnection(DiscoveryNode node, ConnectionProfile c
         }
     }
 
+    /**
+     * Callback invoked when a channel is opened.
+     *
+     * @param channel the opened channel
+     */
+    protected void onChannelOpen(final Channel channel) {
```

Comment:
I am not sure I understand this change. Why do we have to have this onChannelOpen callback? Is it for testing only? I wonder if we can find a different way to do this, it's a no-op in production

Reply:
Yes, it's for testing so we can hook in and quietly close a channel after it's been successfully opened to trigger the issue. I'm open to suggestions on an alternative approach to achieve the same, but this is the cleanest that I found.

Reply:
Yesterday I had tried to use the connection listener to hook in and disconnect; the problem with this is that it was not reliable because it happens on a background thread and would only sporadically succeed to close a channel in time. @s1monw had the clever and sneaky idea to hook into the executor so we can replace it with one that happens synchronously on the same thread and thus we are ensured it happens before the connection finishes. I have pushed this change.